File: src/main/java/leetcode/CombinationSum3.java
package leetcode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Find all valid combinations of k numbers that sum up to n such that the
 * following conditions are true:
 *
 * - Only numbers 1 through 9 are used. - Each number is used at most once.
 *
 * Return a list of all possible valid combinations. The list must not contain
 * the same combination twice, and the combinations may be returned in any
 * order.
 */
final class CombinationSum3 {

    /**
     * Find all valid combinations of k numbers that sum to n.
     *
     * The time complexity O(9^k) as there at most k recursive calls that
     * iterate over at most 9 options.
     *
     * The space complexity is O(k) as there at most k elements in the result.
     *
     * @param k The number of elements.
     * @param n The target sum.
     * @return The combinations of k that sum to n.
     */
    public List<List<Integer>> combinationSum3(final int k, final int n) {

        if (k < 2 || k > 9 || n < 1 || n > 60) {
            return Collections.emptyList();
        }

        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, 1, new ArrayList<>(), k, n);
        return result;
    }

    /**
     * Helper method for backtracking.
     *
     * @param result    The result we are building.
     * @param index     The number from 1-9 we are looking at.
     * @param candidate A candidate for inclusion in the result.
     * @param k         The number of elements.
     * @param n         The target sum.
     */
    private void backtrack(final List<List<Integer>> result, final int index,
            final List<Integer> candidate, final int k, final int n) {

        // we have exceeded n so we discard the candidate
        if (n < 0) {
            return;
        }

        // we have met n and the size is correct so we keep the candidate
        if (n == 0 && candidate.size() == k) {
            result.add(new ArrayList<>(candidate));
            return;
        }

        for (int i = index; i <= 9; i++) {
            // simulate adding i to the next slot
            candidate.add(i);
            // attempt to find remaining slots
            backtrack(result, i + 1, candidate, k, n - i);
            // simulate not adding i to the next slot
            candidate.remove(candidate.size() - 1);
        }
    }
}

File: python/src/combination_sum_3.py
from typing import List


class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        """Find all valid combinations of k numbers that sum up to n such that the following conditions are true:
            - Only numbers 1 through 9 are used.
            - Each number is used at most once.

        Return a list of all possible valid combinations. The list must not contain the same combination twice, and the
        combinations may be returned in any order.

        :param k: The number of combinations that sum to the target sum.
        :param n: The target sum.
        :return List[List[int]: The list of all possible combinations.

        The time complexity is O(k * 9^k) as the recursion depth is k and there are 9 digits to choose from. Faster
        than 82.38% of solutions.

        The space complexity is O(k) for the recursion stack and to hold the list. Less memory than 77.39% of
        solutions.
        """
        self.k = k
        self.n = n
        self.res = []

        self.backtrack(0, [], self.n)

        return self.res

    def backtrack(self, num, cur, target):
        # base cases
        if len(cur) == self.k:
            if target == 0:
                self.res.append(cur.copy())
            return

        for i in range(num + 1, 10):
            if i <= target:
                cur.append(i)
                self.backtrack(i, cur, target - i)
                cur.pop()
            else:
                return
